keybinding concept (#994)
ElektraInitiative/PermaplanT: doc/decisions/frontend_keyhandling.md

# Frontend Keybindings

## Problem

In the development of PermaplanT, we have identified the need for a robust keybinding system to enhance the user experience. Keybindings are utilized in various parts of the application, and we need to establish a cohesive approach for their implementation.

## Constraints

Our keybinding solution should adhere to the following constraints:

- **Ease of Integration**: The keybinding system should be straightforward to integrate into different components of the application.
- **Developer-friendly**: Developers should get an overview where which key binding is assigned.
- **UX**: Keybindings should not lead to unwanted sideeffects.

## Solution

The following questions have been addressed during the conception:

- **How do we configure keybindings?**

- There should be a single source of truth
- it should be easy to understand for developers.

- **Where do we add key listeners?**

- While it's easy to add key listeners to focused components like form inputs, it becomes more challenging when dealing with a component like the Konva map.

> - This is because we cannot directly bind keys to the map layer.

Reviewer: I wouldn't squeeze that in the same decision. It probably doesn't need a decision but just a way that works nicely. (Concept)

> #### Interesting npm package

Reviewer: Please create a subsubsection for each possible solution.

During research following npm package was found. However, currently it is not planned to use it since our custom hook solution is sufficient and easier to adapt.

[react-hotkeys-hook](https://github.com/JohannesKlauss/react-hotkeys-hook/tree/main)

- It provides a hook for binding keys in components and also
- offers scoping abilities.

## Decision

### Keybinding Configuration

> A json file contains all keybindings that are used in the application and serves the following purpose

Reviewer: ```suggestion
A JSON file contains all keybindings that are used in the application and serves the following purpose
```

> - keybindings can easily be changed by editing the file

Reviewer: Consistency within itemize

```suggestion
- Keybindings can easily be changed by editing the file.
```

- Developers have an overview of all keybindings.

Example Structure:

```
{
  "global": {
    "navigateToMap": "Ctrl+m"
  }

>   "planting": {

Reviewer: How can we distinguish layer names, function names etc. here?

>     "exitPlantingMode": "Escape"
>     "search":{
>       "clearSearch": "Escape"

Reviewer: Can we use some standard names? Toolkits usually provide such names.

```suggestion
    "exitPlantingMode": "Esc"
    "search":{
      "clearSearch": "Esc"
```

    }
  }
}
```

- keybindings are scoped.
- there can also be nested scopes
- keys can be combined with '+'

> ### Keyhandling on focused elements

Reviewer: ```suggestion
### Keyhandling on Focused Elements
```

- for components that provide keylisteners and are usually focused on use (e.g. input fields)

> - method: add scope to json, bind a keyhandler and execute action according to configuration

Reviewer: ```suggestion
- method: add scope to JSON, bind a keyhandler and execute action according to configuration
```

- if actions of parent components shouldn't be triggered, event propagation must be stopped

Example

```
    const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
      const action = getActionNameFromKeyEvent("planting.search", event)
      if(action === "clearSearch"){
        resetSearch();
        event.stopPropagation();
      }
    };

> ```

Reviewer: move this concept to guidelines or solution chapter of architecture.

# Frontend Keybindings

## Problem

In the development of PermaplanT, we have identified the need for a robust keybinding system to enhance the user experience. Keybindings are utilized in various parts of the application, and we need to establish a cohesive approach for their implementation.

## Constraints

Our keybinding solution should adhere to the following constraints:

- **Ease of Integration**: The keybinding system should be straightforward to integrate into different components of the application.

> - **User Configuration**: We should plan for future functionality that allows users to configure keybindings according to their preferences.

Reviewer: This is not needed. It should only be developer-friendly: developers should easily get to know where which key binding is assigned.

Having all features to great good UX/UI is however needed (e.g. scoping, maybe @tschawax has more input on that).

> ## Decision / Current Solution

Reviewer: Before Deciding, give solutions (see template)

We have decided to configure keybindings using a JSON structure that includes scopes and associated keybindings.

Example:

```
{
  "global": {
    "Ctrl+M": "navigateToMap"
  }
  "planting": {
    "Escape": "exitPlantingMode"
  }
}
```

> Components that require keybindings can define action handlers for the configured actions. Here's an example of how this can be implemented:

Reviewer: Please one sentence per line to easier comment on individual statements:

```suggestion
Components that require keybindings can define action handlers for the configured actions.
Here's an example of how this can be implemented:
```

Example:

```
  const keyHandlerActions : Record<string, () => void> = {
    exitPlantingMode: () => {
        exitPlantingMode();
    }
  }
```

We can then add key listeners to specific components, and if a key mapping is configured for the pressed key within the associated scope, the corresponding action can be executed.

## Challenges / To be discussed

### Where do we add key listeners?

While it's easy to add key listeners to focused components like form inputs, it becomes more challenging when dealing with a component like the Konva map. This is because we cannot directly bind keys to the map layer.

> Furthermore there are instances where we want to handle keys for map feature even when another component is focused. For example, if a user selects a plant in the right toolbar, pressing the "Escape" key should probably also exit planting mode.

Reviewer: What we want from UI perspective should be described in @tschawax document (#918)

If we also want to clear a search with the "Escape" key, we have to stop event propagation.

> ### Interesting npm package
> 
> During research following npm package was found which seems to be very promising:
> 
> - [react-hotkeys-hook](https://github.com/JohannesKlauss/react-hotkeys-hook/tree/main)
> 
> It provides a hook for binding keys in components and also offers scoping abilities.
> 
> ### Ruled out alternative
> 
> - [react-hotkeys](https://www.npmjs.com/package/react-hotkeys): provides a component that receives a keymap and actionhandlers similar to our planned concept and listens to events if a child of the component is in focus. It seams to be very suitable for encapsulating events, however it doesn't really solve the challenge of focus konva layers and although the librarie is widely used the last realase was 4 years ago.

Reviewer: Put both up to solutions. I agree we can rule out react-hotkeys, as scoping is important from UI perspective.

```suggestion
### React Hotkeys Hook

[react-hotkeys-hook](https://github.com/JohannesKlauss/react-hotkeys-hook/tree/main)

- It provides a hook for binding keys in components and also
- offers scoping abilities.

### React Hotkeys

 [react-hotkeys](https://www.npmjs.com/package/react-hotkeys): provides a component that receives a keymap and actionhandlers similar to our planned concept and listens to events if a child of the component is in focus.

- It seems to be very suitable for encapsulating events, however
- it doesn't really solve the challenge of focus konva layers and
- although the library is widely used the last release was 4 years ago.
```
